File: Embassat/Base/CADArrayDataSource.m
//
//  CADArrayDataSource.m
//  Embassat
//
//  Created by Joan Romano on 25/07/13.
//  Copyright (c) 2013 Crows And Dogs. All rights reserved.
//

#import "CADArrayDataSource.h"

NSString *const CADCellIdentifier = @"CellIdentifier";
NSString *const CADHeaderIdentifier = @"HeaderIdentifier";

@implementation NSObject (CADEMViewModelCollectionDelegate)

- (NSInteger)numberOfSections{return 1;}

@end

@interface CADArrayDataSource ()

@property (nonatomic, strong) id <CADEMViewModelCollectionDelegate> viewModel;
@property (nonatomic, copy) ConfigureViewBlock configureCellBlock;
@property (nonatomic, copy) ConfigureViewBlock configureHeaderBlock;

@end

@implementation CADArrayDataSource

#pragma mark - Init

- (instancetype)initWithViewModel:(id)viewModel
               configureCellBlock:(ConfigureViewBlock)configureBlock
{
    return [self initWithViewModel:viewModel configureCellBlock:configureBlock configureHeaderBlock:NULL];
}

- (instancetype)initWithViewModel:(id<CADEMViewModelCollectionDelegate>)viewModel configureCellBlock:(ConfigureViewBlock)configureBlock configureHeaderBlock:(ConfigureViewBlock)configureHeaderBlock
{
    if (self = [super init])
    {
        _viewModel = viewModel;
        _configureCellBlock = configureBlock;
        _configureHeaderBlock = configureHeaderBlock;
    }
    
    return self;
}

#pragma mark - UICollectionView Data Source

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return [self.viewModel numberOfSections];
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return [self.viewModel numberOfItemsInSection:section];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    id cell = [collectionView dequeueReusableCellWithReuseIdentifier:CADCellIdentifier forIndexPath:indexPath];

    self.configureCellBlock(cell, indexPath);
    
    return cell;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
{
    id view;
    
    view = [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader
                                              withReuseIdentifier:CADHeaderIdentifier
                                                     forIndexPath:indexPath];
    
    self.configureHeaderBlock(view, indexPath);
    
    return view;
}

@end

File: Embassat/Base/CADArrayDataSource.swift
//
//  CADArrayDataSource.swift
//  Embassa't
//
//  Created by Joan Romano on 13/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import UIKit

typealias ConfigureViewBlock = (AnyObject, NSIndexPath) -> Void

public class CADArrayDataSource: NSObject, UICollectionViewDataSource {
    
    static let CADCellIdentifier : String = "CellIdentifier"
    static let CADHeaderIdentifier : String = "HeaderIdentifier"
    
    let viewModel: CADEMViewModelCollectionDelegate
    let configureCellBlock: ConfigureViewBlock
    let configureHeaderBlock: ConfigureViewBlock?
    
    init(viewModel : CADEMViewModelCollectionDelegate, configureCellBlock: ConfigureViewBlock, configureHeaderBlock: ConfigureViewBlock?) {
        self.viewModel = viewModel
        self.configureCellBlock = configureCellBlock
        self.configureHeaderBlock = configureHeaderBlock
    }
    
    public func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        
        if let viewModel = self.viewModel as? NSObject {
            if viewModel.respondsToSelector("numberOfSections") {
                return self.viewModel.numberOfSections!()
            } else {
                return 1
            }
        } else {
            return 1
        }
    }
    
    public func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.viewModel.numberOfItemsInSection(section)
    }
    
    public func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell: AnyObject = collectionView.dequeueReusableCellWithReuseIdentifier(CADArrayDataSource.CADCellIdentifier, forIndexPath: indexPath)
        self.configureCellBlock(cell, indexPath)
        
        return cell as! UICollectionViewCell
    }
    
    public func collectionView(collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, atIndexPath indexPath: NSIndexPath) -> UICollectionReusableView {
        let view: AnyObject = collectionView.dequeueReusableSupplementaryViewOfKind(kind, withReuseIdentifier: CADArrayDataSource.CADHeaderIdentifier, forIndexPath: indexPath)
        if let configureBlock  = self.configureHeaderBlock {
            configureBlock(view, indexPath)
        }
        
        return view as! UICollectionReusableView
    }
}
